packages/babel-parser/src/parser/expression.js

```diff
@@ -2725,4 +2733,163 @@ export default class ExpressionParser extends LValParser {
     this.eat(tt.braceR);
     return this.finishNode<N.ModuleExpression>(node, "ModuleExpression");
   }
+
+  // https://github.com/tc39/proposal-pattern-matching
+  parseMaybeMatchExpression(): N.MatchExpression {
+    const node = this.startNode<N.MatchExpression>();
+    this.next(); // skip "match"
+    this.eat(tt.parenL);
+    node.discriminant = this.parseExpression();
+    this.eat(tt.parenR);
+
+    node.clauses = this.parseMatchClauses();
+    return this.finishNode(node, "MatchExpression");
+  }
+
+  parseMatchClauses(): N.MatchClause[] {
+    this.eat(tt.braceL);
+    const clauses = [];
+    while (!this.eat(tt.braceR)) {
+      if (this.match(tt._if) || this.match(tt._else)) {
+        clauses.push(this.parseConditionalMatchClause());
+      } else if (this.isContextual("when")) {
+        clauses.push(this.parseWhenMatchClause());
+      } else {
+        this.unexpected();
+      }
+    }
+    return clauses;
+  }
+
+  parseConditionalMatchClause(): N.MatchClause {
+    const node = this.startNode<N.MatchClause>();
+    const type = this.state.type;
+    this.next(); // skip "if" / "else"
+    if (type === tt._if) {
+      node.guard = this.parseHeaderExpression();
+    }
+    node.consequent = this.parseStatement("if");
+    return this.finishNode(node, "MatchClause");
+  }
+
+  parseWhenMatchClause(): N.MatchClause {
+    const node = this.startNode<N.MatchClause>();
+    this.next(); // skip "when"
+    this.expect(tt.parenL);
+    node.test = this.parseMaybeBinaryMatchPattern();
+    this.expect(tt.parenR);
+
+    if (this.isLineTerminator()) {
+      this.unexpected(
+        this.state.lastTokEnd,
+        Errors.LineTerminatorAfterMatchPattern,
+      );
+    }
+    node.consequent = this.parseStatement("do");
```

Comment:
```suggestion
    node.consequent = this.parseStatement("when");
```
Guess the context info should be `whene`?

Reply:
the right hand side of a match clause has do expression semantics; i'm not sure what's applicable here.

Reply:
Do `if` and `else` in match context behave like do expressions too?

Reply:
No, if that were possible without breaking the web, then there'd be proposals to make `if` an expression in the first place.